File: training-docs/scripts/generate-docx.js
const fs = require('fs');
const path = require('path');
const {
  Document, Packer, Paragraph, TextRun, Table, TableRow, TableCell,
  Header, Footer, AlignmentType, HeadingLevel, BorderStyle, WidthType,
  ShadingType, PageNumber, PageBreak, TableOfContents, LevelFormat
} = require('docx');

// Read all markdown files in order
const docsDir = path.join(__dirname, 'docs');
const chapterDirs = [
  '00-intro', '01-hardware', '02-connection', '03-tools', '04-protocols',
  '05-internals', '06-rtos-linux', '07-programming', '08-ai', '09-pcb',
  '10-mechanical', '11-project-design', '12-modules'
];

const CHAPTER_NAMES = {
  '00-intro': '第〇篇：认识单片机',
  '01-hardware': '第一篇：硬件基础',
  '02-connection': '第二篇：电脑如何连接单片机',
  '03-tools': '第三篇：开发工具与工程结构',
  '04-protocols': '第四篇：通信协议',
  '05-internals': '第五篇：单片机内部机制',
  '06-rtos-linux': '第六篇：RTOS 与 Linux',
  '07-programming': '第七篇：编程实践',
  '08-ai': '第八篇：AI 编程',
  '09-pcb': '第九篇：PCB 设计',
  '10-mechanical': '第十篇：机械结构',
  '11-project-design': '第十一篇：项目方案设计',
  '12-modules': '第十二篇：常见模块与实战',
};

function readAllFiles() {
  const allContent = [];
  for (const dir of chapterDirs) {
    const dirPath = path.join(docsDir, dir);
    if (!fs.existsSync(dirPath)) continue;
    const files = fs.readdirSync(dirPath).filter(f => f.endsWith('.md')).sort();
    if (files.length === 0) continue;

    allContent.push({ type: 'chapter-title', text: CHAPTER_NAMES[dir], file: dir });

    for (const file of files) {
      const filePath = path.join(dirPath, file);
      const content = fs.readFileSync(filePath, 'utf-8');
      allContent.push({ type: 'file', text: content, file: filePath });
    }
  }
  return allContent;
}

// Parse a single markdown file into blocks
function parseMarkdown(text) {
  const lines = text.split('\n');
  const blocks = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    // Code block
    if (line.trim().startsWith('```')) {
      const lang = line.trim().slice(3).trim();
      const codeLines = [];
      i++;
      while (i < lines.length && !lines[i].trim().startsWith('```')) {
        codeLines.push(lines[i]);
        i++;
      }
      i++; // skip closing ```
      blocks.push({ type: 'code', lang, content: codeLines.join('\n') });
      continue;
    }

    // Table
    if (line.includes('|') && line.trim().startsWith('|')) {
      const tableRows = [];
      while (i < lines.length && lines[i].includes('|') && lines[i].trim().startsWith('|')) {
        const cells = lines[i].split('|').filter(c => c.trim() !== '');
        tableRows.push(cells.map(c => c.trim()));
        i++;
      }
      // Filter out separator rows (e.g., |---|---|)
      const filteredRows = tableRows.filter(row => !row.every(c => /^[-:]+$/.test(c)));
      if (filteredRows.length > 0) {
        blocks.push({ type: 'table', rows: filteredRows });
      }
      continue;
    }

    // Callout (::: tip/warning/danger)
    if (/^:::\s*(tip|warning|danger|info)/.test(line.trim())) {
      const calloutType = line.trim().match(/^:::\s*(tip|warning|danger|info)/)[1];
      const calloutLines = [];
      i++;
      while (i < lines.length && !lines[i].trim().startsWith(':::')) {
        calloutLines.push(lines[i]);
        i++;
      }
      i++; // skip closing :::
      blocks.push({ type: 'callout', calloutType, content: calloutLines.join('\n') });
      continue;
    }

    // Heading
    if (/^#{1,6}\s/.test(line)) {
      const match = line.match(/^(#{1,6})\s+(.+)/);
      blocks.push({ type: 'heading', level: match[1].length, text: match[2] });
      i++;
      continue;
    }

    // Empty line
    if (line.trim() === '') {
      i++;
      continue;
    }

    // Normal paragraph - collect until empty line or special block
    const paraLines = [];
    while (i < lines.length && lines[i].trim() !== ''
           && !lines[i].trim().startsWith('```')
           && !(lines[i].includes('|') && lines[i].trim().startsWith('|'))
           && !/^:::\s*(tip|warning|danger|info)/.test(lines[i].trim())
           && !/^#{1,6}\s/.test(lines[i])) {
      paraLines.push(lines[i]);
      i++;
    }
    if (paraLines.length > 0) {
      blocks.push({ type: 'paragraph', content: paraLines.join('\n') });
    }
  }
  return blocks;
}

// Parse inline markdown in text (bold, code, links)
function parseInline(text) {
  if (!text) return [];
  const runs = [];
  let remaining = text;

  while (remaining.length > 0) {
    // Bold
    let match = remaining.match(/^\*\*(.+?)\*\*/);
    if (match) {
      runs.push(new TextRun({ text: match[1], bold: true }));
      remaining = remaining.slice(match[0].length);
      continue;
    }
    // Inline code
    match = remaining.match(/^`([^`]+)`/);
    if (match) {
      runs.push(new TextRun({ text: match[1], font: "Consolas", size: 20, shading: { fill: "E8E8E8", type: ShadingType.CLEAR } }));
      remaining = remaining.slice(match[0].length);
      continue;
    }
    // Link [text](url)
    match = remaining.match(/^\[([^\]]+)\]\(([^)]+)\)/);
    if (match) {
      runs.push(new TextRun({ text: match[1], color: "0066CC", underline: {} }));
      remaining = remaining.slice(match[0].length);
      continue;
    }
    // Plain text
    const nextSpecial = remaining.search(/`|\*\*|\[/);
    if (nextSpecial === -1) {
      runs.push(new TextRun({ text: remaining }));
      remaining = '';
    } else {
      runs.push(new TextRun({ text: remaining.slice(0, nextSpecial) }));
      remaining = remaining.slice(nextSpecial);
    }
  }
  return runs.filter(r => r && r.options && r.options.text && r.options.text.length > 0);
}

// Callout colors
const CALLOUT_COLORS = {
  tip: { bg: "D5F5E3", border: "27AE60", label: "💡 提示" },
  warning: { bg: "FDEBD0", border: "E67E22", label: "⚠️ 注意" },
  danger: { bg: "FADBD8", border: "E74C3C", label: "🚫 警告" },
  info: { bg: "D6EAF8", border: "2980B9", label: "ℹ️ 信息" },
};

const border = { style: BorderStyle.SINGLE, size: 1, color: "CCCCCC" };
const borders = { top: border, bottom: border, left: border, right: border };

function createDocxElements(blocks) {
  const elements = [];

  for (const block of blocks) {
    if (block.type === 'heading') {
      const level = Math.min(block.level, 3);
      const headingMap = {
        1: HeadingLevel.HEADING_1,
        2: HeadingLevel.HEADING_2,
        3: HeadingLevel.HEADING_3,
      };
      elements.push(new Paragraph({
        heading: headingMap[level],
        children: [new TextRun({ text: block.text })],
      }));
    } else if (block.type === 'paragraph') {
      const runs = parseInline(block.content);
      if (runs.length > 0) {
        elements.push(new Paragraph({
          children: runs,
          spacing: { after: 120, line: 360 },
        }));
      }
    } else if (block.type === 'code') {
      const codeLines = block.content.split('\n');
      codeLines.forEach(line => {
        elements.push(new Paragraph({
          children: [new TextRun({ text: line || ' ', font: "Consolas", size: 18 })],
          shading: { fill: "F5F5F5", type: ShadingType.CLEAR },
          spacing: { before: 0, after: 0, line: 280 },
          indent: { left: 360 },
        }));
      });
      // Add a small gap after code block
      elements.push(new Paragraph({ spacing: { after: 120 }, children: [] }));
    } else if (block.type === 'table') {
      const headerRow = block.rows[0];
      const dataRows = block.rows.length > 1 ? block.rows.slice(1) : [];
      const colWidth = Math.floor(9026 / headerRow.length); // A4 content width

      elements.push(new Table({
        width: { size: 9026, type: WidthType.DXA },
        columnWidths: headerRow.map(() => colWidth),
        rows: [
          new TableRow({
            children: headerRow.map(cell =>
              new TableCell({
                borders,
                width: { size: colWidth, type: WidthType.DXA },
                shading: { fill: "D5E8F0", type: ShadingType.CLEAR },
                margins: { top: 60, bottom: 60, left: 100, right: 100 },
                children: [new Paragraph({
                  children: [new TextRun({ text: cell, bold: true, size: 20 })]
                })]
              })
            )
          }),
          ...dataRows.map(row =>
            new TableRow({
              children: row.map(cell =>
                new TableCell({
                  borders,
                  width: { size: colWidth, type: WidthType.DXA },
                  margins: { top: 60, bottom: 60, left: 100, right: 100 },
                  children: [new Paragraph({
                    children: parseInline(cell),
                    spacing: { after: 0 },
                  })]
                })
              )
            })
          )
        ]
      }));
      elements.push(new Paragraph({ spacing: { after: 120 }, children: [] }));
    } else if (block.type === 'callout') {
      const colors = CALLOUT_COLORS[block.calloutType] || CALLOUT_COLORS.info;
      const calloutLines = block.content.split('\n');

      // Label line
      elements.push(new Paragraph({
        children: [new TextRun({ text: colors.label, bold: true, size: 20 })],
        shading: { fill: colors.bg, type: ShadingType.CLEAR },
        border: { left: { style: BorderStyle.SINGLE, size: 8, color: colors.border } },
        spacing: { before: 160, after: 0 },
        indent: { left: 200 },
      }));
      // Content lines
      calloutLines.forEach(line => {
        if (line.trim()) {
          elements.push(new Paragraph({
            children: parseInline(line),
            shading: { fill: colors.bg, type: ShadingType.CLEAR },
            border: { left: { style: BorderStyle.SINGLE, size: 8, color: colors.border } },
            spacing: { before: 0, after: 0 },
            indent: { left: 200 },
          }));
        }
      });
      elements.push(new Paragraph({ spacing: { after: 120 }, children: [] }));
    }
  }
  return elements;
}

async function main() {
  console.log("Reading markdown files...");
  const allContent = readAllFiles();

  const docElements = [];
  let firstFile = true;

  for (const item of allContent) {
    if (item.type === 'chapter-title') {
      // Add page break before each chapter (except first)
      if (!firstFile) {
        docElements.push(new Paragraph({ children: [new PageBreak()] }));
      }
      firstFile = false;

      docElements.push(new Paragraph({
        heading: HeadingLevel.HEADING_1,
        children: [new TextRun({ text: item.text, size: 36 })],
        spacing: { before: 360, after: 240 },
        alignment: AlignmentType.CENTER,
      }));
    } else {
      const blocks = parseMarkdown(item.text);
      const elements = createDocxElements(blocks);
      docElements.push(...elements);
    }
  }

  // Create document
  const doc = new Document({
    styles: {
      default: {
        document: {
          run: { font: "Microsoft YaHei", size: 22 }
        }
      },
      paragraphStyles: [
        {
          id: "Heading1", name: "Heading 1", basedOn: "Normal", next: "Normal", quickFormat: true,
          run: { size: 32, bold: true, font: "Microsoft YaHei" },
          paragraph: { spacing: { before: 240, after: 240 }, outlineLevel: 0 }
        },
        {
          id: "Heading2", name: "Heading 2", basedOn: "Normal", next: "Normal", quickFormat: true,
          run: { size: 28, bold: true, font: "Microsoft YaHei" },
          paragraph: { spacing: { before: 180, after: 180 }, outlineLevel: 1 }
        },
        {
          id: "Heading3", name: "Heading 3", basedOn: "Normal", next: "Normal", quickFormat: true,
          run: { size: 24, bold: true, font: "Microsoft YaHei" },
          paragraph: { spacing: { before: 120, after: 120 }, outlineLevel: 2 }
        },
      ]
    },
    sections: [
      // Cover page
      {
        properties: {
          page: {
            size: { width: 11906, height: 16838 },
            margin: { top: 1440, right: 1440, bottom: 1440, left: 1440 }
          }
        },
        children: [
          new Paragraph({ spacing: { before: 4800 }, children: [] }),
          new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [new TextRun({ text: "电赛入门指南", size: 56, bold: true, font: "Microsoft YaHei" })],
          }),
          new Paragraph({ spacing: { before: 200 }, children: [] }),
          new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [new TextRun({ text: "从零开始学嵌入式", size: 36, color: "555555" })],
          }),
          new Paragraph({ spacing: { before: 400 }, children: [] }),
          new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [new TextRun({ text: "2026 江苏省电子设计竞赛备赛教程", size: 26, color: "777777" })],
          }),
          new Paragraph({ spacing: { before: 600 }, children: [] }),
          new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [new TextRun({ text: "适用平台：MSPM0G3507 / STM32F103", size: 22, color: "999999" })],
          }),
        ]
      },
      // TOC page
      {
        properties: {
          page: {
            size: { width: 11906, height: 16838 },
            margin: { top: 1440, right: 1440, bottom: 1440, left: 1440 }
          }
        },
        headers: {
          default: new Header({
            children: [new Paragraph({
              alignment: AlignmentType.RIGHT,
              children: [new TextRun({ text: "电赛入门指南", size: 18, color: "999999" })],
            })]
          })
        },
        footers: {
          default: new Footer({
            children: [new Paragraph({
              alignment: AlignmentType.CENTER,
              children: [new TextRun({ text: "第 ", size: 18 }), new TextRun({ children: [PageNumber.CURRENT], size: 18 }), new TextRun({ text: " 页", size: 18 })],
            })]
          })
        },
        children: [
          new Paragraph({
            heading: HeadingLevel.HEADING_1,
            children: [new TextRun({ text: "目录" })],
          }),
          new TableOfContents("目录", { hyperlink: true, headingStyleRange: "1-3" }),
        ]
      },
      // Main content
      {
        properties: {
          page: {
            size: { width: 11906, height: 16838 },
            margin: { top: 1440, right: 1440, bottom: 1440, left: 1440 }
          }
        },
        headers: {
          default: new Header({
            children: [new Paragraph({
              alignment: AlignmentType.RIGHT,
              children: [new TextRun({ text: "电赛入门指南", size: 18, color: "999999" })],
            })]
          })
        },
        footers: {
          default: new Footer({
            children: [new Paragraph({
              alignment: AlignmentType.CENTER,
              children: [new TextRun({ text: "第 ", size: 18 }), new TextRun({ children: [PageNumber.CURRENT], size: 18 }), new TextRun({ text: " 页", size: 18 })],
            })]
          })
        },
        children: docElements,
      }
    ]
  });

  const outputPath = path.join(__dirname, '..', '电赛入门指南.docx');
  console.log(`Generating Word document to ${outputPath}...`);
  const buffer = await Packer.toBuffer(doc);
  fs.writeFileSync(outputPath, buffer);
  console.log(`Done! File saved to ${outputPath}`);
  console.log(`File size: ${(buffer.length / 1024 / 1024).toFixed(1)} MB`);
}

main().catch(console.error);

File: training-docs/scripts/generate-docx.py
"""Generate Word document from all markdown files."""
import os
import re
from docx import Document
from docx.shared import Pt, Inches, Cm, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.style import WD_STYLE_TYPE
from docx.oxml.ns import qn, nsdecls
from docx.oxml import parse_xml
import docx.oxml

docs_dir = os.path.join(os.path.dirname(__file__), 'docs')
output_path = os.path.join(os.path.dirname(__file__), '..', '电赛入门指南.docx')

CHAPTER_DIRS = [
    '00-intro', '01-hardware', '02-connection', '03-tools', '04-protocols',
    '05-internals', '06-rtos-linux', '07-programming', '08-ai', '09-pcb',
    '10-mechanical', '11-project-design', '12-modules'
]

CHAPTER_NAMES = {
    '00-intro': '第〇篇：认识单片机',
    '01-hardware': '第一篇：硬件基础',
    '02-connection': '第二篇：电脑如何连接单片机',
    '03-tools': '第三篇：开发工具与工程结构',
    '04-protocols': '第四篇：通信协议',
    '05-internals': '第五篇：单片机内部机制',
    '06-rtos-linux': '第六篇：RTOS 与 Linux',
    '07-programming': '第七篇：编程实践',
    '08-ai': '第八篇：AI 编程',
    '09-pcb': '第九篇：PCB 设计',
    '10-mechanical': '第十篇：机械结构',
    '11-project-design': '第十一篇：项目方案设计',
    '12-modules': '第十二篇：常见模块与实战',
}

CALLOUT_COLORS = {
    'tip': ('D5F5E3', '27AE60', '💡 提示'),
    'warning': ('FDEBD0', 'E67E22', '⚠️ 注意'),
    'danger': ('FADBD8', 'E74C3C', '🚫 警告'),
    'info': ('D6EAF8', '2980B9', 'ℹ️ 信息'),
}


def set_cell_shading(cell, color):
    """Set cell background color."""
    shading_elm = parse_xml(f'<w:shd {nsdecls("w")} w:fill="{color}" w:val="clear"/>')
    cell._tc.get_or_add_tcPr().append(shading_elm)


def add_code_block(doc, code_text):
    """Add a code block with gray background and Consolas font."""
    for line in code_text.split('\n'):
        p = doc.add_paragraph()
        p.style = doc.styles['Code']
        run = p.add_run(line if line else ' ')
        run.font.name = 'Consolas'
        run.font.size = Pt(9)
        run._element.rPr.rFonts.set(qn('w:eastAsia'), 'Consolas')


def add_paragraph_with_inline(doc, text):
    """Add a paragraph with inline formatting (bold, code, links)."""
    p = doc.add_paragraph()
    p.style = doc.styles['Body Text']
    # Simple regex-based inline parsing
    remaining = text
    while remaining:
        # Bold
        m = re.match(r'\*\*(.+?)\*\*', remaining)
        if m:
            run = p.add_run(m.group(1))
            run.bold = True
            remaining = remaining[m.end():]
            continue
        # Inline code
        m = re.match(r'`([^`]+)`', remaining)
        if m:
            run = p.add_run(m.group(1))
            run.font.name = 'Consolas'
            run.font.size = Pt(9)
            run.font.shading = parse_xml(f'<w:shd {nsdecls("w")} w:fill="E8E8E8" w:val="clear"/>')
            remaining = remaining[m.end():]
            continue
        # Link
        m = re.match(r'\[([^\]]+)\]\(([^)]+)\)', remaining)
        if m:
            run = p.add_run(m.group(1))
            run.font.color.rgb = RGBColor(0, 102, 204)
            run.underline = True
            remaining = remaining[m.end():]
            continue
        # Plain text
        next_special = re.search(r'`|\*\*|\[', remaining)
        if not next_special:
            p.add_run(remaining)
            remaining = ''
        else:
            p.add_run(remaining[:next_special.start()])
            remaining = remaining[next_special.start():]


def add_table(doc, rows):
    """Add a table to the document."""
    num_cols = len(rows[0])
    table = doc.add_table(rows=len(rows), cols=num_cols)
    table.style = 'Table Grid'

    for i, row_data in enumerate(rows):
        for j, cell_text in enumerate(row_data):
            cell = table.cell(i, j)
            cell.text = ''
            p = cell.paragraphs[0]
            run = p.add_run(cell_text)
            run.font.size = Pt(9)
            if i == 0:
                run.bold = True
                set_cell_shading(cell, 'D5E8F0')

    doc.add_paragraph()  # spacing after table


def add_callout(doc, callout_type, text):
    """Add a callout block with colored left border and background."""
    colors = CALLOUT_COLORS.get(callout_type, CALLOUT_COLORS['info'])
    bg, border_color, label = colors

    # Label
    p = doc.add_paragraph()
    run = p.add_run(label)
    run.bold = True
    run.font.size = Pt(10)
    set_paragraph_border(p, border_color)
    set_paragraph_shading(p, bg)

    # Content
    for line in text.split('\n'):
        if line.strip():
            p = doc.add_paragraph()
            run = p.add_run(line)
            run.font.size = Pt(10)
            set_paragraph_border(p, border_color)
            set_paragraph_shading(p, bg)

    doc.add_paragraph()  # spacing


def set_paragraph_border(paragraph, color):
    """Add left border to paragraph."""
    pPr = paragraph._p.get_or_add_pPr()
    pBdr = parse_xml(
        f'<w:pBdr {nsdecls("w")}>'
        f'<w:left w:val="single" w:sz="8" w:space="4" w:color="{color}"/>'
        f'</w:pBdr>'
    )
    pPr.append(pBdr)


def set_paragraph_shading(paragraph, color):
    """Add background shading to paragraph."""
    pPr = paragraph._p.get_or_add_pPr()
    shd = parse_xml(f'<w:shd {nsdecls("w")} w:fill="{color}" w:val="clear"/>')
    pPr.append(shd)


def parse_markdown(text):
    """Parse markdown text into blocks."""
    lines = text.split('\n')
    blocks = []
    i = 0

    while i < len(lines):
        line = lines[i]

        # Code block
        if line.strip().startswith('```'):
            lang = line.strip()[3:].strip()
            code_lines = []
            i += 1
            while i < len(lines) and not lines[i].strip().startswith('```'):
                code_lines.append(lines[i])
                i += 1
            i += 1  # skip closing ```
            blocks.append(('code', '\n'.join(code_lines)))
            continue

        # Table
        if '|' in line and line.strip().startswith('|'):
            table_rows = []
            while i < len(lines) and '|' in lines[i] and lines[i].strip().startswith('|'):
                cells = [c.strip() for c in lines[i].split('|') if c.strip()]
                table_rows.append(cells)
                i += 1
            # Filter separator rows
            filtered = [r for r in table_rows if not all(re.match(r'^[-:]+$', c) for c in r)]
            if filtered:
                blocks.append(('table', filtered))
            continue

        # Callout
        m = re.match(r'^:::\s*(tip|warning|danger|info)', line.strip())
        if m:
            callout_type = m.group(1)
            callout_lines = []
            i += 1
            while i < len(lines) and not lines[i].strip().startswith(':::'):
                callout_lines.append(lines[i])
                i += 1
            i += 1  # skip closing :::
            blocks.append(('callout', callout_type, '\n'.join(callout_lines)))
            continue

        # Heading
        m = re.match(r'^(#{1,6})\s+(.+)', line)
        if m:
            level = len(m.group(1))
            blocks.append(('heading', min(level, 3), m.group(2)))
            i += 1
            continue

        # Empty line
        if line.strip() == '':
            i += 1
            continue

        # Normal paragraph
        para_lines = []
        while (i < len(lines) and lines[i].strip() != ''
               and not lines[i].strip().startswith('```')
               and not (lines[i].strip().startswith('|') and '|' in lines[i])
               and not re.match(r'^:::\s*(tip|warning|danger|info)', lines[i].strip())
               and not re.match(r'^#{1,6}\s', lines[i])):
            para_lines.append(lines[i])
            i += 1
        if para_lines:
            blocks.append(('paragraph', '\n'.join(para_lines)))

    return blocks


def setup_styles(doc):
    """Configure document styles."""
    # Normal style
    style = doc.styles['Normal']
    style.font.name = 'Microsoft YaHei'
    style.font.size = Pt(11)
    style.paragraph_format.line_spacing = 1.5
    style.paragraph_format.space_after = Pt(6)
    style.element.rPr.rFonts.set(qn('w:eastAsia'), 'Microsoft YaHei')

    # Body Text
    style = doc.styles['Body Text']
    style.font.name = 'Microsoft YaHei'
    style.font.size = Pt(11)
    style.paragraph_format.line_spacing = 1.5

    # Code style
    style = doc.styles.add_style('Code', WD_STYLE_TYPE.PARAGRAPH)
    style.font.name = 'Consolas'
    style.font.size = Pt(9)
    style.paragraph_format.space_before = Pt(0)
    style.paragraph_format.space_after = Pt(0)
    style.paragraph_format.line_spacing = 1.1
    # Gray background
    pPr = style.element.get_or_add_pPr()
    shd = parse_xml(f'<w:shd {nsdecls("w")} w:fill="F5F5F5" w:val="clear"/>')
    pPr.append(shd)

    # Heading styles
    for level, (size, name) in enumerate([(22, 'Heading 1'), (18, 'Heading 2'), (14, 'Heading 3')], 1):
        style = doc.styles[name]
        style.font.name = 'Microsoft YaHei'
        style.font.size = Pt(size)
        style.font.bold = True
        style.font.color.rgb = RGBColor(0, 0, 0)
        style.element.rPr.rFonts.set(qn('w:eastAsia'), 'Microsoft YaHei')


def add_cover_page(doc):
    """Add cover page."""
    for _ in range(8):
        doc.add_paragraph()

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run('电赛入门指南')
    run.font.size = Pt(36)
    run.bold = True
    run.font.name = 'Microsoft YaHei'
    run._element.rPr.rFonts.set(qn('w:eastAsia'), 'Microsoft YaHei')

    doc.add_paragraph()

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run('从零开始学嵌入式')
    run.font.size = Pt(22)
    run.font.color.rgb = RGBColor(85, 85, 85)

    doc.add_paragraph()
    doc.add_paragraph()

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run('2026 江苏省电子设计竞赛备赛教程')
    run.font.size = Pt(16)
    run.font.color.rgb = RGBColor(119, 119, 119)

    doc.add_paragraph()
    doc.add_paragraph()

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run('适用平台：MSPM0G3507 / STM32F103')
    run.font.size = Pt(12)
    run.font.color.rgb = RGBColor(153, 153, 153)

    doc.add_page_break()


def main():
    print("Creating Word document...")
    doc = Document()
    setup_styles(doc)

    # Add cover page
    add_cover_page(doc)

    # Add TOC placeholder
    p = doc.add_paragraph('目录', style='Heading 1')
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    doc.add_paragraph()

    # Add TOC field
    paragraph = doc.add_paragraph()
    run = paragraph.add_run()
    fldChar = parse_xml(f'<w:fldChar {nsdecls("w")} w:fldCharType="begin"/>')
    run._r.append(fldChar)
    run = paragraph.add_run()
    instrText = parse_xml(f'<w:instrText {nsdecls("w")} xml:space="preserve"> TOC \\o "1-3" \\h \\z \\u </w:instrText>')
    run._r.append(instrText)
    run = paragraph.add_run()
    fldChar2 = parse_xml(f'<w:fldChar {nsdecls("w")} w:fldCharType="separate"/>')
    run._r.append(fldChar2)
    run = paragraph.add_run('（请在 Word 中右键此处，选择"更新域"以生成目录）')
    run.font.color.rgb = RGBColor(153, 153, 153)
    run.font.size = Pt(10)
    run = paragraph.add_run()
    fldChar3 = parse_xml(f'<w:fldChar {nsdecls("w")} w:fldCharType="end"/>')
    run._r.append(fldChar3)

    doc.add_page_break()

    # Process all chapters
    total_files = 0
    for chapter_dir in CHAPTER_DIRS:
        dir_path = os.path.join(docs_dir, chapter_dir)
        if not os.path.isdir(dir_path):
            continue

        # Add chapter title
        chapter_name = CHAPTER_NAMES.get(chapter_dir, chapter_dir)
        doc.add_heading(chapter_name, level=1)

        # Process files
        files = sorted([f for f in os.listdir(dir_path) if f.endswith('.md')])
        for filename in files:
            filepath = os.path.join(dir_path, filename)
            with open(filepath, 'r', encoding='utf-8') as f:
                text = f.read()

            blocks = parse_markdown(text)
            for block in blocks:
                if block[0] == 'heading':
                    doc.add_heading(block[2], level=block[1])
                elif block[0] == 'paragraph':
                    add_paragraph_with_inline(doc, block[1])
                elif block[0] == 'code':
                    add_code_block(doc, block[1])
                    doc.add_paragraph()  # spacing
                elif block[0] == 'table':
                    add_table(doc, block[1])
                elif block[0] == 'callout':
                    add_callout(doc, block[1], block[2])

            total_files += 1
            if total_files % 10 == 0:
                print(f"  Processed {total_files} files...")

    # Add headers and footers to all sections
    for section in doc.sections:
        section.top_margin = Cm(2.54)
        section.bottom_margin = Cm(2.54)
        section.left_margin = Cm(2.54)
        section.right_margin = Cm(2.54)

        # Header
        header = section.header
        header.is_linked_to_previous = False
        p = header.paragraphs[0]
        p.alignment = WD_ALIGN_PARAGRAPH.RIGHT
        run = p.add_run('电赛入门指南')
        run.font.size = Pt(9)
        run.font.color.rgb = RGBColor(153, 153, 153)

        # Footer with page number
        footer = section.footer
        footer.is_linked_to_previous = False
        p = footer.paragraphs[0]
        p.alignment = WD_ALIGN_PARAGRAPH.CENTER
        run = p.add_run('第 ')
        run.font.size = Pt(9)
        # Page number field
        run = p.add_run()
        fldChar1 = parse_xml(f'<w:fldChar {nsdecls("w")} w:fldCharType="begin"/>')
        run._r.append(fldChar1)
        run = p.add_run()
        instrText = parse_xml(f'<w:instrText {nsdecls("w")} xml:space="preserve"> PAGE </w:instrText>')
        run._r.append(instrText)
        run = p.add_run()
        fldChar2 = parse_xml(f'<w:fldChar {nsdecls("w")} w:fldCharType="separate"/>')
        run._r.append(fldChar2)
        run = p.add_run('1')
        run.font.size = Pt(9)
        run = p.add_run()
        fldChar3 = parse_xml(f'<w:fldChar {nsdecls("w")} w:fldCharType="end"/>')
        run._r.append(fldChar3)
        run = p.add_run(' 页')
        run.font.size = Pt(9)

    print(f"Saving to {output_path}...")
    doc.save(output_path)

    size_mb = os.path.getsize(output_path) / (1024 * 1024)
    print(f"Done! File: {output_path} ({size_mb:.1f} MB)")
    print(f"Total files processed: {total_files}")


if __name__ == '__main__':
    main()
